src/proxy/inbound.rs

```diff
@@ -259,138 +291,173 @@ impl Inbound {
-                    parse_baggage_header(req.headers().get_all(BAGGAGE_HEADER)).unwrap_or_default();
-
-                let source = match from_gateway {
-                    true => None, // we cannot lookup source workload since we don't know the network, see https://github.com/istio/ztunnel/issues/515
-                    false => {
-                        let src_network_addr = NetworkAddress {
-                            // we can assume source network is our network because we did not traverse a gateway
-                            network: conn.dst_network.to_string(),
-                            address: source_ip,
-                        };
-                        // Find source info. We can lookup by XDS or from connection attributes
-                        state.fetch_workload(&src_network_addr).await
-                    }
-                };
+        // Orig has 15008, swap with the real port
+        let conn = Connection {
+            dst: if sandwich {
+                SocketAddr::new(conn.dst.ip(), 15088)
+            } else {
+                addr
+            },
+            ..conn
+        };
 
-                let derived_source = metrics::DerivedWorkload {
-                    identity: conn.src_identity.clone(),
-                    cluster_id: baggage.cluster_id,
-                    namespace: baggage.namespace,
-                    workload_name: baggage.workload_name,
-                    revision: baggage.revision,
-                    ..Default::default()
-                };
-                let ds = proxy::guess_inbound_service(&conn, upstream_service, &upstream);
-                let connection_metrics = ConnectionOpen {
-                    reporter: Reporter::destination,
-                    source,
-                    derived_source: Some(derived_source),
-                    destination: Some(upstream),
-                    connection_security_policy: metrics::SecurityPolicy::mutual_tls,
-                    destination_service: ds,
-                };
-                let status_code = match Self::handle_inbound(
-                    Hbone(req),
-                    enable_original_source.then_some(source_ip),
-                    addr,
-                    metrics,
-                    connection_metrics,
-                    None,
-                    socket_factory.as_ref(),
-                )
-                .in_current_span()
-                .await
-                {
-                    Ok(_) => StatusCode::OK,
-                    Err(_) => StatusCode::SERVICE_UNAVAILABLE,
-                };
+        let dst_network_addr = NetworkAddress {
+            network: conn.dst_network.to_string(), // dst must be on our network
+            address: addr.ip(),
+        };
+
+        // for sandwich, we haven't resolved the VIP
+        // otherwise, this is a workload and should have the service
+        let wl_svcs = state.fetch_workload_services(&dst_network_addr).await;
```

Comment:
for sandwich we should really look up the service using this VIP and look up the waypoint workload and make sure they're related